WIP: Fix use of the macOS SDK

chrisburr opened the pull request:
Changes in  this PR:

* Force `conda-build` to use a shorter install prefix (exposes PCH relocation bugs).
* Remove `clangdev` from the run time dependencies.
* **[osx]** Make the shell error message clearer if `CONDA_BUILD_SYSROOT` isn't set.
* **[osx]** Prevent cling from adding the macOS SDK path to `libCling`. (a5ef92b)
* **[osx]** Distribute `etc/allDict.cxx.pch` as it can't be generated without a working compiler.
* **[osx]** Disable prefix replacement for `etc/allDict.cxx.pch` as both text and binary mode fail. I think the only effect is causing error messages to the wrong paths (as in LCG).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

henryiii commented on recipe/activate.fish:
```diff
@@ -7,8 +7,11 @@ switch (uname)
+                echo "WARNING: Compiling likely won't work unless you: download the macOS 10.9 SDK, set CONDA_BUILD_SYSROOT and reactivate the environment."
+                echo "You can probably ignore this warning and just omit + or ++ when executing ROOT macros."
+            else
+                export SDKROOT="$CONDA_BUILD_SYSROOT"
```
`set -x SDKROOT "$CONDA_BUILD_SYSROOT"`

henryiii commented on recipe/deactivate.sh:
```diff
@@ -1,3 +1,13 @@
+    # Only if not in the base env (let's be nice)
+    if [ "${CONDA_DEFAULT_ENV}"  != "base" ] ; then
+        if [ ! -z "${CONDA_BUILD_SYSROOT}" ] && [ -z "${SDKROOT}" ] ; then
+            unset SDKROOT
```
This could unset a user set var. Better to use a marker variable when setting to check for.

henryiii commented on .azure-pipelines/azure-pipelines-osx.yml:
```diff
@@ -48,6 +48,8 @@ jobs:
       set -x -e
       conda install -n base -c conda-forge --quiet --yes conda-forge-ci-setup=2 conda-build
+      # Patch conda use a shorter testing prefix
+      pip install --upgrade git+https://github.com/chrisburr/conda-build.git@patch-1
```
Why this? Just curious.

chrisburr replied:
This picks up my PR to `conda-build`: https://github.com/conda/conda-build/pull/3386

It uses a 78 character prefix when testing instead of 80 characters (which is used for building). This exposes bugs in my (failed) attempts to edit the paths in the PCH.

henryiii replied:
Will you include this in the merged version?

chrisburr replied:
I see little reason not to, it will be removed next time the feedstock is rendered.

henryiii replied:
True, okay then!

henryiii commented:
I should be able to start helping again tomorrow.

chrisburr commented:
I've added a description of what I've changed. I've also removed the changes to `SDKROOT` as I _think_ they aren't needed.

If the latest commit passes I think it's ready to merge.

henryiii commented on recipe/meta.yaml:
```diff
@@ -19,15 +19,19 @@ source:
     - patches/root-6.16.02_allow_setting_clang_version.patch  # [osx]
     - patches/root-x.xx.xx_python_mac_os.patch  # [osx]
     - patches/root-x.xx.xx-copy-abspath-headers.patch  # [osx]
+    - patches/root-x.xx.xx-osx-remove-hardcoded-sysroot.patch  # [osx]
```
Should this been made to ROOT?

chrisburr replied:
I wasn't planning to but a CMake option could be added though I'm not sure it would ever be useful for anything other than conda.

henryiii replied:
Okay. Might be worth a bit of discussion, but it's probably mot worth of an option. Most systems I can think of would have a static location of the current sdk and build separately for each version. Maybe gentoo prefix, but I think even that has per macOS version builds.

henryiii commented:
Okay, looks like a good solution.

chrisburr commented:
As it will take a long time to deploy and the current version is broken I'm going to merge this so there is a (hopefully) working package in the morning.
